File: ProjectTest/OneLang/CSharp/One/Ast/AstTypes.cs
using One.Ast;

namespace One.Ast
{
    public interface IHasTypeArguments {
        Type_[] typeArguments { get; set; }
    }
    
    public interface IInterfaceType : IType {
        Type_[] typeArguments { get; set; }
        
        IInterface getDecl();
    }
    
    public class Type_ : IType {
        public static bool isGeneric(Type_ type) {
            if (type is GenericsType)
                return true;
            else if (type is ClassType classType)
                return classType.typeArguments.some(x => Type_.isGeneric(x));
            else if (type is InterfaceType intType)
                return intType.typeArguments.some(x => Type_.isGeneric(x));
            else if (type is LambdaType lambdType)
                return lambdType.parameters.some(x => Type_.isGeneric(x.type)) || Type_.isGeneric(lambdType.returnType);
            else
                return false;
        }
        
        public static bool equals(Type_ type1, Type_ type2) {
            if (type1 == null || type2 == null)
                throw new Error("Type is missing!");
            if (type1 is VoidType && type2 is VoidType)
                return true;
            if (type1 is AnyType && type2 is AnyType)
                return true;
            if (type1 is GenericsType genType && type2 is GenericsType genType2)
                return genType.typeVarName == genType2.typeVarName;
            if (type1 is EnumType enumType && type2 is EnumType enumType2)
                return enumType.decl == enumType2.decl;
            if (type1 is LambdaType lambdType2 && type2 is LambdaType lambdType3)
                return Type_.equals(lambdType2.returnType, lambdType3.returnType) && lambdType2.parameters.length() == lambdType3.parameters.length() && lambdType2.parameters.every((t, i) => Type_.equals(t.type, lambdType3.parameters.get(i).type));
            if (type1 is ClassType classType2 && type2 is ClassType classType3)
                return classType2.decl == classType3.decl && classType2.typeArguments.length() == classType3.typeArguments.length() && classType2.typeArguments.every((t, i) => Type_.equals(t, classType3.typeArguments.get(i)));
            if (type1 is InterfaceType intType2 && type2 is InterfaceType intType3)
                return intType2.decl == intType3.decl && intType2.typeArguments.length() == intType3.typeArguments.length() && intType2.typeArguments.every((t, i) => Type_.equals(t, intType3.typeArguments.get(i)));
            return false;
        }
        
        public static bool isAssignableTo(Type_ toBeAssigned, Type_ whereTo) {
            // AnyType can assigned to any type except to void
            if (toBeAssigned is AnyType && !(whereTo is VoidType))
                return true;
            // any type can assigned to AnyType except void
            if (whereTo is AnyType && !(toBeAssigned is VoidType))
                return true;
            // any type can assigned to GenericsType except void
            if (whereTo is GenericsType && !(toBeAssigned is VoidType))
                return true;
            // null can be assigned anywhere
            // TODO: filter out number and boolean types...
            if (toBeAssigned is NullType && !(whereTo is VoidType))
                return true;
            
            if (Type_.equals(toBeAssigned, whereTo))
                return true;
            
            if (toBeAssigned is ClassType classType4 && whereTo is ClassType classType5)
                return (classType4.decl.baseClass != null && Type_.isAssignableTo(classType4.decl.baseClass, classType5)) || classType4.decl == classType5.decl && classType4.typeArguments.every((x, i) => Type_.isAssignableTo(x, classType5.typeArguments.get(i)));
            if (toBeAssigned is ClassType classType6 && whereTo is InterfaceType intType4)
                return (classType6.decl.baseClass != null && Type_.isAssignableTo(classType6.decl.baseClass, intType4)) || classType6.decl.baseInterfaces.some(x => Type_.isAssignableTo(x, intType4));
            if (toBeAssigned is InterfaceType intType5 && whereTo is InterfaceType intType6)
                return intType5.decl.baseInterfaces.some(x => Type_.isAssignableTo(x, intType6)) || intType5.decl == intType6.decl && intType5.typeArguments.every((x, i) => Type_.isAssignableTo(x, intType6.typeArguments.get(i)));
            if (toBeAssigned is LambdaType lambdType4 && whereTo is LambdaType lambdType5)
                return lambdType4.parameters.length() == lambdType5.parameters.length() && lambdType4.parameters.every((p, i) => Type_.isAssignableTo(p.type, lambdType5.parameters.get(i).type)) && (Type_.isAssignableTo(lambdType4.returnType, lambdType5.returnType) || lambdType5.returnType is GenericsType);
            
            return false;
        }
        
        public virtual string repr() {
            return "U:UNKNOWN";
        }
    }
    
    public class TypeHelper {
        public static string argsRepr(Type_[] args) {
            return args.length() == 0 ? "" : $"<{args.map(x => x.repr()).join(", ")}>";
        }
    }
    
    public class PrimitiveType : Type_ {
        
    }
    
    public class VoidType : PrimitiveType {
        public static VoidType instance;
        
        static VoidType()
        {
            VoidType.instance = new VoidType();
        }
        
        public override string repr() {
            return "Void";
        }
    }
    
    public class AnyType : PrimitiveType {
        public static AnyType instance;
        
        static AnyType()
        {
            AnyType.instance = new AnyType();
        }
        
        public override string repr() {
            return "Any";
        }
    }
    
    public class NullType : PrimitiveType {
        public static NullType instance;
        
        static NullType()
        {
            NullType.instance = new NullType();
        }
        
        public override string repr() {
            return "Null";
        }
    }
    
    public class GenericsType : Type_ {
        public string typeVarName;
        
        public GenericsType(string typeVarName): base()
        {
            this.typeVarName = typeVarName;
        }
        
        public override string repr() {
            return $"G:{this.typeVarName}";
        }
    }
    
    public class EnumType : Type_ {
        public Enum_ decl;
        
        public EnumType(Enum_ decl): base()
        {
            this.decl = decl;
        }
        
        public override string repr() {
            return $"E:{this.decl.name}";
        }
    }
    
    public class InterfaceType : Type_, IHasTypeArguments, IInterfaceType {
        public Interface decl;
        public Type_[] typeArguments { get; set; }
        
        public InterfaceType(Interface decl, Type_[] typeArguments): base()
        {
            this.decl = decl;
            this.typeArguments = typeArguments;
        }
        
        public IInterface getDecl() {
            return this.decl;
        }
        
        public override string repr() {
            return $"I:{this.decl.name}{TypeHelper.argsRepr(this.typeArguments)}";
        }
    }
    
    public class ClassType : Type_, IHasTypeArguments, IInterfaceType {
        public Class decl;
        public Type_[] typeArguments { get; set; }
        
        public ClassType(Class decl, Type_[] typeArguments): base()
        {
            this.decl = decl;
            this.typeArguments = typeArguments;
        }
        
        public IInterface getDecl() {
            return this.decl;
        }
        
        public override string repr() {
            return $"C:{this.decl.name}{TypeHelper.argsRepr(this.typeArguments)}";
        }
    }
    
    public class UnresolvedType : Type_, IHasTypeArguments {
        public string typeName;
        public Type_[] typeArguments { get; set; }
        
        public UnresolvedType(string typeName, Type_[] typeArguments): base()
        {
            this.typeName = typeName;
            this.typeArguments = typeArguments;
        }
        
        public override string repr() {
            return $"X:{this.typeName}{TypeHelper.argsRepr(this.typeArguments)}";
        }
    }
    
    public class LambdaType : Type_ {
        public MethodParameter[] parameters;
        public Type_ returnType;
        
        public LambdaType(MethodParameter[] parameters, Type_ returnType): base()
        {
            this.parameters = parameters;
            this.returnType = returnType;
            if (returnType == null) { }
        }
        
        public override string repr() {
            return $"L:({this.parameters.map(x => x.type.repr()).join(", ")})=>{this.returnType.repr()}";
        }
    }
}

File: ProjectTest/OneLang/Python/OneLang/One/Ast/AstTypes.py
from OneLangStdLib import *
import OneLangStdLib as one
import OneLang.One.Ast.Interfaces as ints

class Type:
    def __init__(self):
        pass
    
    @classmethod
    def is_generic(cls, type):
        if isinstance(type, GenericsType):
            return True
        elif isinstance(type, ClassType):
            return ArrayHelper.some(lambda x: Type.is_generic(x), type.type_arguments)
        elif isinstance(type, InterfaceType):
            return ArrayHelper.some(lambda x: Type.is_generic(x), type.type_arguments)
        elif isinstance(type, LambdaType):
            return ArrayHelper.some(lambda x: Type.is_generic(x.type), type.parameters) or Type.is_generic(type.return_type)
        else:
            return False
    
    @classmethod
    def equals(cls, type1, type2):
        if type1 == None or type2 == None:
            raise Error("Type is missing!")
        if isinstance(type1, VoidType) and isinstance(type2, VoidType):
            return True
        if isinstance(type1, AnyType) and isinstance(type2, AnyType):
            return True
        if isinstance(type1, GenericsType) and isinstance(type2, GenericsType):
            return type1.type_var_name == type2.type_var_name
        if isinstance(type1, EnumType) and isinstance(type2, EnumType):
            return type1.decl == type2.decl
        if isinstance(type1, LambdaType) and isinstance(type2, LambdaType):
            return Type.equals(type1.return_type, type2.return_type) and len(type1.parameters) == len(type2.parameters) and ArrayHelper.every(lambda t, i: Type.equals(t.type, type2.parameters[i].type), type1.parameters)
        if isinstance(type1, ClassType) and isinstance(type2, ClassType):
            return type1.decl == type2.decl and len(type1.type_arguments) == len(type2.type_arguments) and ArrayHelper.every(lambda t, i: Type.equals(t, type2.type_arguments[i]), type1.type_arguments)
        if isinstance(type1, InterfaceType) and isinstance(type2, InterfaceType):
            return type1.decl == type2.decl and len(type1.type_arguments) == len(type2.type_arguments) and ArrayHelper.every(lambda t, i: Type.equals(t, type2.type_arguments[i]), type1.type_arguments)
        return False
    
    @classmethod
    def is_assignable_to(cls, to_be_assigned, where_to):
        # AnyType can assigned to any type except to void
        if isinstance(to_be_assigned, AnyType) and not (isinstance(where_to, VoidType)):
            return True
        # any type can assigned to AnyType except void
        if isinstance(where_to, AnyType) and not (isinstance(to_be_assigned, VoidType)):
            return True
        # any type can assigned to GenericsType except void
        if isinstance(where_to, GenericsType) and not (isinstance(to_be_assigned, VoidType)):
            return True
        # null can be assigned anywhere
        # TODO: filter out number and boolean types...
        if isinstance(to_be_assigned, NullType) and not (isinstance(where_to, VoidType)):
            return True
        
        if Type.equals(to_be_assigned, where_to):
            return True
        
        if isinstance(to_be_assigned, ClassType) and isinstance(where_to, ClassType):
            return (to_be_assigned.decl.base_class != None and Type.is_assignable_to(to_be_assigned.decl.base_class, where_to)) or to_be_assigned.decl == where_to.decl and ArrayHelper.every(lambda x, i: Type.is_assignable_to(x, where_to.type_arguments[i]), to_be_assigned.type_arguments)
        if isinstance(to_be_assigned, ClassType) and isinstance(where_to, InterfaceType):
            return (to_be_assigned.decl.base_class != None and Type.is_assignable_to(to_be_assigned.decl.base_class, where_to)) or ArrayHelper.some(lambda x: Type.is_assignable_to(x, where_to), to_be_assigned.decl.base_interfaces)
        if isinstance(to_be_assigned, InterfaceType) and isinstance(where_to, InterfaceType):
            return ArrayHelper.some(lambda x: Type.is_assignable_to(x, where_to), to_be_assigned.decl.base_interfaces) or to_be_assigned.decl == where_to.decl and ArrayHelper.every(lambda x, i: Type.is_assignable_to(x, where_to.type_arguments[i]), to_be_assigned.type_arguments)
        if isinstance(to_be_assigned, LambdaType) and isinstance(where_to, LambdaType):
            return len(to_be_assigned.parameters) == len(where_to.parameters) and ArrayHelper.every(lambda p, i: Type.is_assignable_to(p.type, where_to.parameters[i].type), to_be_assigned.parameters) and (Type.is_assignable_to(to_be_assigned.return_type, where_to.return_type) or isinstance(where_to.return_type, GenericsType))
        
        return False
    
    def repr(self):
        return "U:UNKNOWN"

class TypeHelper:
    def __init__(self):
        pass
    
    @classmethod
    def args_repr(cls, args):
        return "" if len(args) == 0 else f'''<{", ".join(list(map(lambda x: x.repr(), args)))}>'''

class PrimitiveType(Type):
    def __init__(self):
        super().__init__()

@one.static_init
class VoidType(PrimitiveType):
    @classmethod
    def static_init(cls):
        cls.instance = cls()
    
    def __init__(self):
        super().__init__()
    
    def repr(self):
        return "Void"

@one.static_init
class AnyType(PrimitiveType):
    @classmethod
    def static_init(cls):
        cls.instance = cls()
    
    def __init__(self):
        super().__init__()
    
    def repr(self):
        return "Any"

@one.static_init
class NullType(PrimitiveType):
    @classmethod
    def static_init(cls):
        cls.instance = cls()
    
    def __init__(self):
        super().__init__()
    
    def repr(self):
        return "Null"

class GenericsType(Type):
    def __init__(self, type_var_name):
        self.type_var_name = type_var_name
        super().__init__()
    
    def repr(self):
        return f'''G:{self.type_var_name}'''

class EnumType(Type):
    def __init__(self, decl):
        self.decl = decl
        super().__init__()
    
    def repr(self):
        return f'''E:{self.decl.name}'''

class InterfaceType(Type):
    def __init__(self, decl, type_arguments):
        self.decl = decl
        self.type_arguments = type_arguments
        super().__init__()
    
    def get_decl(self):
        return self.decl
    
    def repr(self):
        return f'''I:{self.decl.name}{TypeHelper.args_repr(self.type_arguments)}'''

class ClassType(Type):
    def __init__(self, decl, type_arguments):
        self.decl = decl
        self.type_arguments = type_arguments
        super().__init__()
    
    def get_decl(self):
        return self.decl
    
    def repr(self):
        return f'''C:{self.decl.name}{TypeHelper.args_repr(self.type_arguments)}'''

class UnresolvedType(Type):
    def __init__(self, type_name, type_arguments):
        self.type_name = type_name
        self.type_arguments = type_arguments
        super().__init__()
    
    def repr(self):
        return f'''X:{self.type_name}{TypeHelper.args_repr(self.type_arguments)}'''

class LambdaType(Type):
    def __init__(self, parameters, return_type):
        self.parameters = parameters
        self.return_type = return_type
        super().__init__()
        if return_type == None:
            pass
    
    def repr(self):
        return f'''L:({", ".join(list(map(lambda x: x.type.repr(), self.parameters)))})=>{self.return_type.repr()}'''